include/ghex/bulk_communication_object.hpp

```diff
@@ -354,27 +384,41 @@ class bulk_communication_object
             // check if field has the right domain
             if (f.domain_id() == p.domain_id())
             {
+                auto m_it = m_tag_map.insert(std::pair<int,int>(f.domain_id(),0)).first;
                 // loop over halos and set up source ranges
                 for (auto h_it = p.send_halos().begin(); h_it != p.send_halos().end(); ++h_it)
                 {
+                    int q=0;
                     for (auto it = h_it->second.rbegin(); it != h_it->second.rend(); ++it)
                     {
                         const auto& c = *it;
                         s_range.m_ranges.back().emplace_back(
-                            m_comm, f, c, h_it->first.mpi_rank, h_it->first.tag); 
+                            m_comm, f, c, h_it->first.mpi_rank
+                            , (m_it->second + h_it->first.tag+1)*10000 + q
+                            // alternatively rely on message ordering:
```

Comment:
If this is useful in some cases, shouldn't we guard this with a macro?

Reply:
Maybe. I will add a macro.